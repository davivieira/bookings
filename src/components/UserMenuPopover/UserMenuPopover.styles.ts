import { Box, Button, styled } from "@mui/material";

export const SCPopoverContainer = styled(Box)(() => ({
  display: 'flex',
  flexDirection: 'column',
  gap: 10,
  paddingTop: 10,
  paddingBottom: 10
}))

export const SCPopoverOption = styled(Button)(() => ({
  width: '100%'
}))