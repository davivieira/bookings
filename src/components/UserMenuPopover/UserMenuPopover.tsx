import { useNavigate } from 'react-router-dom'
import { SCPopoverContainer, SCPopoverOption } from './UserMenuPopover.styles'
import { useDispatch } from 'react-redux';
import { resetUser } from '@/features/userSlice';

function UserMenuPopover() {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const handleNavigate = () => {
    navigate('/bookings');
  }

  const handleLogout = () => {
    dispatch(resetUser())
    sessionStorage.removeItem('user');
    navigate('/');
  }

  return (
    <SCPopoverContainer>
      <SCPopoverOption onClick={handleNavigate}>My Bookings</SCPopoverOption>
      <SCPopoverOption onClick={handleLogout}>Logout</SCPopoverOption>
    </SCPopoverContainer>
  )
}

export default UserMenuPopover