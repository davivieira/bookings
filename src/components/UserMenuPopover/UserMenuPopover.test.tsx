import { render, screen } from '@testing-library/react';
import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import UserMenuPopover from './UserMenuPopover';
import userEvent from '@testing-library/user-event';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('UserMenuPopover', () => {

  beforeEach(() => {
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as unknown as jest.Mock).mockReturnValue(null);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders user menu popover with options', () => {
    render(<UserMenuPopover />);

    expect(screen.getByText(/My Bookings/i)).toBeInTheDocument();
    expect(screen.getByText(/Logout/i)).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<UserMenuPopover />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('navigates to bookings page when "My Bookings" is clicked', async () => {
    const user = userEvent.setup();
    render(<UserMenuPopover />);

    await user.click(screen.getByText(/My Bookings/i));
    expect(mockNavigate).toHaveBeenCalledWith('/bookings');
  });

  it('navigates to home page when "logout" is clicked', async () => {
    const user = userEvent.setup();
    render(<UserMenuPopover />);

    await user.click(screen.getByText(/logout/i));
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
});
