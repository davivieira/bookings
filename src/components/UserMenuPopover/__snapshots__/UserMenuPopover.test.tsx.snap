// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenuPopover matches snapshot 1`] = `
<div
  class="MuiBox-root css-5tl4eu"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-mf6l92-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    My Bookings
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-mf6l92-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Logout
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
